crates/provider/src/fillers/nonce.rs

```diff
@@ -92,33 +94,57 @@ impl NonceFiller {
         N: Network,
         T: Transport + Clone,
     {
-        // locks dashmap internally for a short duration to clone the `Arc`
-        let mutex = Arc::clone(self.nonces.entry(from).or_default().value());
-
-        // locks the value (does not lock dashmap)
-        let mut nonce = mutex.lock().await;
-        match *nonce {
-            Some(ref mut nonce) => {
-                *nonce += 1;
-                Ok(*nonce)
-            }
-            None => {
-                // initialize the nonce if we haven't seen this account before
-                let initial_nonce = provider.get_transaction_count(from).await?;
-                *nonce = Some(initial_nonce);
-                Ok(initial_nonce)
-            }
+        // Use `u64::MAX` as a sentinel value to indicate that the nonce has not been fetched yet.
+        const NONE: u64 = u64::MAX;
+
+        // Locks dashmap internally for a short duration to clone the `Arc`.
+        // We also don't want to hold the dashmap lock through the await point below.
+        let nonce = Arc::clone(
+            self.nonces.entry(from).or_insert_with(|| Arc::new(AtomicU64::new(NONE))).value(),
+        );
+
+        let prev_nonce = nonce.fetch_add(1, Ordering::SeqCst);
+        let current_nonce;
+        if prev_nonce == NONE {
+            // Initialize the nonce if we haven't seen this account before.
+            current_nonce = provider.get_transaction_count(from).await?;
+            nonce.store(current_nonce, Ordering::SeqCst);
```

Comment:
this is now reachable by two concurrent calls that can return the same value.

I guess we need a mutex at least for the initial fetch?